### ISSUE TITLE
Create occur buffer in background on minibuffer exit

### ISSUE THREAD

**Issue by clemera** [oantolin/embark#85]
When discussing [selectrum-repeat](https://github.com/raxod502/selectrum/issues/319) it occurred to me that it may be nice to have an automatic occur buffer created on minibuffer exit which would basically replace the functionality of such a command. This way you could always go back to your last results. What do you think?

**Comment by minad** [comment]
It would also be the perfect reason to why I wouldn't want to implement a `consult-resume` command. See also https://github.com/minad/consult/issues/6#issuecomment-753729831.

**Comment by oantolin** [comment]
Does `(add-hook 'minibuffer-exit-hook 'embark-occur)` work or does the hook run a little to late? I'll test later when I'm at the computer.

**Comment by minad** [comment]
Oh, and the slowdown is annoying if you implement it that way, since the marginalia annotators are executed always. In order to support this properly, only the un-annotated candidate list should be stored and then copied to an occur buffer and annotated by `embark-resurrect`. It seems this is not a feature we can get for free. But I still think Embark is the right place for this. Certainly better than having a `selectrum/consult-repeat/resume` command!

**Comment by oantolin** [comment]
The occur buffer could be requested in grid view, which runs no annotator, and then be put in the correct initial view for the type when resurrected.

**Comment by oantolin** [comment]
We could also just store the information needed to make the buffer and not actually make it until it is time to ressurect. That's slightly more work, but probably better.

**Comment by minad** [comment]
> The occur buffer could be requested in grid view, which runs no annotator, and then be put in the correct initial view for the type when resurrected.

Right!

> We could also just store the information needed to make the buffer and not actually make it until it is time to ressurect. That's slightly more work, but probably better.

This is what I would like! I would prefer the most efficient way to implement this, since this will be executed every time and in almost all cases it will be executed unnecessarily. If the resurrect feature would result in any noticeable slowdown, I probably wouldn't want to have this.

**Comment by oantolin** [comment]
Also, instead of storing the candidates and later making an occur buffer, we could store the minibuffer input and rerun the command. That feels more like what Ivy resume does.

**Comment by oantolin** [comment]
The version we discussed first would be an embark-occur after the fact, the last thing I proposed would be Embark-become after the fact, becoming the same command again.

**Comment by oantolin** [comment]
Maybe it even makes sense to have both.

**Comment by minad** [comment]
> How about the name embark-occurred?

I like to have a different name for faster completion when typing it. Furthermore I think more distinct names are better.

> Also, instead of storing the candidates and later making an occur buffer, we could store the minibuffer input and rerun the command. That feels more like what Ivy resume does.

Okay, this works too, and maybe this is even better, since you can always get an occur buffer later. This is basically ivy-resume for free? I would then vote to have this!

**Comment by oantolin** [comment]
> I like to have a different name for faster completion when typing it. Furthermore I think more distinct names are better.

**Comment by clemera** [comment]
> Also, instead of storing the candidates and later making an occur buffer, we could store the minibuffer input and rerun the command.

Hm, this sounds more difficult as a command could prompt you multiple times, how would that be handled?

**Comment by minad** [comment]
> Hm, this sounds more difficult as a command could prompt you multiple times, how would that be handled?

I think it is more important to handle the common case of one completion. Does the default action even work if you have multiple completions?

**Comment by oantolin** [comment]
> > Also, instead of storing the candidates and later making an occur buffer, we could store the minibuffer input and rerun the command.
> 
> Hm, this sounds more difficult as a command could prompt you multiple times, how would that be handled?

By ignoring the issue! This is what `embark-default-action` and `embark-become` already do and I haven't received any complaints. 😆

**Comment by minad** [comment]
> By ignoring the issue! This is what embark-default-action and embark-become already do and I haven't received any complaints. laughing

This is exactly how I would have "solved" this too! :+1:

**Comment by oantolin** [comment]
`embark-default-action` means "take this completion from the nth time the command prompted you, ignore the value of n, rerun the same command, and inject the target at the first prompt".

**Comment by minad** [comment]
Yes, and by doing exactly the same we get `embark-resume` for free. I like this very much! It is a hack but it will work in most cases.  Having multiple subsequent completion sessions is not common. One more reason to have consult-async/consult-grep implemented as is :)

**Comment by clemera** [comment]
Not multiple completions sessions but `read-string` and then `completing-read` or stuff like that. But I'm okay with this, too maybe not worth to complicate it.

**Comment by minad** [comment]
But we could also consider solving this for real, somehow recording the sequence of completing-read selections and injecting them in order. It wouldn't be difficult to do this I guess. But this discussion is orthogonal to the `embark-resume` discussion. If we get `embark-resume` based on default action it will be fine for now and if we decide to fix the multi completing-read thing too, it will automatically work!

**Comment by oantolin** [comment]
Good. I'll implement the cheap version later today. I think I'll call it `embark-repeat` rather than resume, because the built in commands are called `repeat` and `repeat-complex-command`

**Comment by clemera** [comment]
Sounds great, thank you!

**Comment by oantolin** [comment]
That was easy. It works, but maybe it needs a little design work. 

- Maybe keep a ring of the last 10 commands and inputs?

- This records every command that uses the minibuffer, including, for example M-x. Maybe some commands should be skipped.

Play around with it, let me know what you think.

**Comment by clemera** [comment]
When repeating a session where you accepted the candidate the input isn't restored but the last submitted candidate.  Is the idea of saving the candidates and restoring an occur buffer from it abandoned? I liked the idea that you just would need to switch to the buffer (no additional keybinding to remember). If it is to slow maybe the buffer could be created via an idle timer so it would be hard to notice and the exit wouldn't be slowed down. 

When the minibuffer session gets restored there might also be the problem that there is additional state (filtering method) that isn't restored which was the reason for the Selectrum issue I linked to. Maybe we could also combine both and the session could be restored from the occur buffer? Maybe this could even be a general feature for occur buffers?

> Maybe keep a ring of the last 10 commands and inputs?

That sounds cool, could be useful from time to time.

> This records every command that uses the minibuffer, including, for example M-x. Maybe some commands should be skipped.

Not sure about that, the purpose for me would be to automatically remember the last thing I did and go back to it.

**Comment by kljohann** [comment]
One minor nit: When Emacs has just started the "last-command-and-input" is nil and the current implementation tries to `(command-execute nil)`. This is due to the following peculiarity of `pcase-let`:
> Each EXP should match (i.e. be of compatible structure) to its
respective PATTERN; a mismatch may signal an error or may go
undetected, binding variables to arbitrary values, such as nil.

I think there are two sensible fixes:
```elisp
  ;; either emit a verbose message and exit:
  (unless embark--last-command-and-input
    (user-error "No last command"))
  (pcase-let ...)

  ;; or silently do nothing by e.g. replacing pcase-let:
  (when-let ((cmd (car embark--last-command-and-input))
             (input (cdr embark--last-command-and-input)))
  ...)
```

**Comment by minad** [comment]
> When repeating a session where you accepted the candidate the input isn't restored but the last submitted candidate. 

Yes, this would be better! @oantolin Any chance to make this work? This certainly needs some special casing! Using just the selected candidate is clearly not what is needed for a proper `embark-repeat`. See also what I write below regarding history management!

> Is the idea of saving the candidates and restoring an occur buffer from it abandoned? I liked the idea that you just would need to switch to the buffer (no additional keybinding to remember). If it is to slow maybe the buffer could be created via an idle timer so it would be hard to notice and the exit wouldn't be slowed down.

No, I wouldn't like this. This sounds like an ugly hack for something simple (instead only the candidates should be stored quickly as a list). But I am in favor of abandoning this idea and instead focus on `embark-repeat` with the input!

>> Maybe keep a ring of the last 10 commands and inputs?

> That sounds cool, could be useful from time to time.

Somehow I have the feeling that this is starting to overlap with the input history functionality, https://github.com/raxod502/selectrum/pull/185 and https://github.com/minad/consult/issues/137. I think we should get back to the drawing board instead of rushing to a solution. Maybe this whole history management should go into Embark?

**Comment by clemera** [comment]
I also noticed the overlap of the problem to get the input but I don't feel the history management would belong into embark.

**Comment by clemera** [comment]
> But I am in favor of abandoning this idea and instead focus on embark-repeat with the input!

What about the problem of additional session state like filtering method. What I like more about the buffer version is that saving and restoring the candidate list would at least assure you get the same candidates.

**Comment by minad** [comment]
> I also noticed the overlap of the problem to get the input but I don't feel the history management would belong into embark.

Yes, maybe reconsider the idea of making it another package. The overlap that I am seeing is the following: If we have a proper input history, all Embark needs to do is take the last element from the input history and inject it as input without a RET to the last recorded command. Then we get `embark-repeat` as we want it. Right now the history is always the selected candidate by default, so we cannot just use the history. Another thing is that @oantolin is proposing another ring of commands+last input, sounds also like a history to me, but a very special one. Maybe I am overthinking this and maybe I am messing up orthogonal things here.

All I want is that we do not rush to half baked solutions, putting history functionality overly quickly into Embark, Consult or Selectrum. Note that the input history has been proposed originally for Selectrum, then we discussed about moving it to Consult. Then consult-repeat has been proposed and we thought that it fits better in Embark etc etc. It seems to be not completely trivial to find a good solution here.

>> But I am in favor of abandoning this idea and instead focus on embark-repeat with the input!
> What about the problem of additional session state like filtering method. What I like more about the buffer version is that saving and restoring the candidate list would at least assure you get the same candidates.

Yes sure this problem exists too. But I am more in favor of in-band information these days, as we did for the async filters (with grep regexp, filter string, command arguments). And as I am using when I specify the filter method via a special character in orderless. So maybe the idea of additional hidden state should be given up and instead we should focus only on good support for the in-band info. Note that the history also does not store the hidden state (related https://github.com/raxod502/selectrum/issues/319). I think making this hidden state thing work is not a good idea, it works against how Emacs histories work.

**Comment by clemera** [comment]
Reusing the the input history for this sounds like a good idea! A separate package would do but maybe the occur buffer is abandoned you could reconsider adding it to consult?

**Comment by minad** [comment]
> Reusing the the input history for this sounds like a good idea! A separate package would do but when the whole occur buffer idea is abandoned you could reconsider adding it to consult? Experimental for now and if all works well there could be a separate package.

The missing piece is the input injection and this is provided by Embark. But it is not much code so we could also do that without Embark. But maybe this should just be extra package. It has a narrow scope, but it is different from how both Consult and Embark operate as of now. I personally feel that it is closer to Embark, since Embark already does injection and some tracking and it enhances existing commands. Consult is not doing something like this, it does not globally modify commands. The only thing which Consult is doing as of now is that it messes with the history for Consult command, e.g. consult-line replaces the history element with the input. Well and there are commands which inspect the histories, so there is certainly some relation again.

The feature list:

* Tracking input history (text property of the normal history items or separate input histories?)
* Toggling between input/candidate histories
* Tracking latest commands and latest inputs
* Provide a session repeat command which restarts the latest commands and latest inputs

**Comment by clemera** [comment]
Thanks for nailing it down! And as you lined out the latest inputs can be retrieved from the input histories. The tricky part is how to receive and save the input histories. Usually completion frameworks expose a variable which stores the last input like I mentioned in the selectrum issue which could be used to handle that, I don't know how this would be done for icomplete/default completion, there you would probably need advice?

**Comment by minad** [comment]
Yes, I have to check the minibuffer contents - Already doing that in consult! 

I think I will make a prototype and then we decide if we shall put it into Embark or Consult.

**Comment by minad** [comment]
Here is some basic prototype. It seems it would be completely isolated from Consult and Embark and could go to a separate package if we feel like it. It is a bit the same situation as with Marginalia, since it enhances existing commands.

~~~ elisp
;;; -*- lexical-binding: t -*-

(require 'subr-x)

;; TODO these must be stored using savehist
(defvar chronicle--input-histories nil)
(defvar chronicle--command-history nil)

(defvar-local chronicle--last-input nil)
(defvar-local chronicle--last-command nil)
(defvar-local chronicle--last-history nil)

(defun chronicle--record-input ()
  (setq chronicle--last-input (minibuffer-contents-no-properties)))

(defun chronicle--minibuffer-setup ()
  ;; TODO there must be a possibility to configure commands
  ;; such that they prefer the input history by default.
  ;; Then we should toggle the histories here.
  ;; And I should remove the :input history functionality from Consult.
  (setq chronicle--last-command this-command)
  (add-hook 'post-command-hook #'chronicle--record-input nil t))

(defun chronicle--minibuffer-exit ()
  (unless (eq t minibuffer-history-variable)
    (let* ((hist minibuffer-history-variable)
           (cand (minibuffer-contents-no-properties))
           (cmd chronicle--last-command)
           (input chronicle--last-input))
      (run-at-time
       0 nil
       (lambda ()
         (let ((hist-val (symbol-value hist)))
           ;; Try to detect if minibuffer has been aborted.
           ;; TODO Is there a better way?
           ;; TODO Change this such that aborted sessions can also be repeated?
           (when (string= (car hist-val) cand)
             (push (cons cmd input) chronicle--command-history)
             (if-let (input-history (assoc hist chronicle--input-histories))
                 ;; TODO Respect history-delete-duplicates
                 ;; TODO Limit history size
                 (setcdr input-history (cons input (cdr input-history)))
               (push (cons hist (list input)) chronicle--input-histories)))))))))

(defun chronicle-repeat (elem)
  (interactive
   (list
    (if-let (elem (car chronicle--command-history))
        elem
      (error "Command history is empty"))))
  (minibuffer-with-setup-hook
      (lambda () (insert (cdr elem)))
    (command-execute (car elem))))

(defun chronicle-select ()
  (interactive)
  (let ((cands (mapcar (lambda (elem)
                         (cons (format "%s %s" (car elem) (cdr elem)) elem))
                       chronicle--command-history)))
  (chronicle-repeat (cdr (assoc (completing-read "History: " cands nil t nil t) cands)))))

(defun chronicle-toggle ()
  (interactive)
  ;; TODO toggle minibuffer-history-variable such
  ;; that we can access both input and original history
  ;; TODO this needs some toggling back, when exiting the minibuffer...
  )

(add-hook 'minibuffer-setup-hook #'chronicle--minibuffer-setup)
(add-hook 'minibuffer-exit-hook #'chronicle--minibuffer-exit)
~~~

If we go that route and implement the functionality as proposed here, I would remove the input history support from consult. Furthermore `embark-repeat` should be removed (revert https://github.com/oantolin/embark/commit/622f6eabfce4e3f15a34774646f2a9325b6c852f).

Opinions?

**Comment by oantolin** [comment]
Wow, differing timezones means there's lots of stuff for me to read now! 😄

**Comment by oantolin** [comment]
@clemera:

> When repeating a session where you accepted the candidate the input isn't restored but the last submitted candidate

Damn, I didn't even notice! Technically the candidate *is* the last input, depending on how you exit. For example, `minibuffer-force-complete-and-exit` modifies the minibuffer contents so it includes the full candidate and then exits. You are right that instead of the last input we should store "the last minibuffer contents before `RET`".

>> This records every command that uses the minibuffer, including, for example M-x. Maybe some commands should be skipped.

> Not sure about that, the purpose for me would be to automatically remember the last thing I did and go back to it.

I just meant that if you do `M-x some-command` and `some-command` does not use the minibuffer, then the command that used the minibuffer is `execute-extended-command`, while you might want that one to "not count".

@kljohann I noticed the issue with `pcase-let`, it's annoying. I like your error message version.

**Comment by clemera** [comment]
Looks like a good path to take, just saving the input in post-command-hook is a great idea, and this works for selectrum, too :) I'm also for removing `selectrum-repeat` when the alternative is available.

**Comment by clemera** [comment]
@oantolin 
> I just meant that if you do M-x some-command and some-command does not use the minibuffer, then the command that used the minibuffer is execute-extended-command, while you might want that one to "not count".

I see, could be nice to have it as an configureable option

**Comment by oantolin** [comment]
> Yes, maybe reconsider the idea of making it another package. The overlap that I am seeing is the following: If we have a proper input history, all Embark needs to do is take the last element from the input history and inject it as input without a RET to the last recorded command.

Yes, a separate package sounds better. (And I like the name `chronicle` you suggest later!) This package would in charge of *defining* what "last input" even means! I thought it meant the minibuffer contents as you are exiting, but that's not it, since, depending on what completion UI your are using and what you have `RET` bound to, often the selected candidate *is* the last minibuffer contents contents. Figuring out what we mean by last input sounds like the job of a separate package, albeit a small one.

>  Another thing is that @oantolin is proposing another ring of commands+last input, sounds also like a history to me, but a very special one. Maybe I am overthinking this and maybe I am messing up orthogonal things here.

Yes, I'd say it is yet another type of history.

> All I want is that we do not rush to half baked solutions, putting history functionality overly quickly into Embark, Consult or Selectrum. Note that the input history has been proposed originally for Selectrum, then we discussed about moving it to Consult. Then consult-repeat has been proposed for Consult and we thought that it fits better in Embark etc. It seems to be not completely trivial to find a good solution here.

Agreed! I will remove `embark-repeat` in favor of discussing this more and reaching a fully fledged design.

> Yes sure this problem exists too. But I am more in favor of in-band information these days [...] So maybe the idea of additional hidden state should be given up and instead we should focus only on good support for the in-band info. Note that the history also does not store the hidden state [...]. I think making this hidden state thing work is not a good idea, it works against how Emacs histories work.

Agreed!

**Comment by oantolin** [comment]
> The missing piece is the input injection and this is provided by Embark. But it is not much code so we could also do that without Embark. But maybe this should just be extra package. It has a narrow scope, but it is different from how both Consult and Embark operate as of now.

Agreed. The injection is really is not much code at all, see `embark-repeat`. Dealing with injection is not enough reason to think this belongs in Embark. I don't oppose adding an appropriate definition of "last input" to Embark, and doing this there, but I'm starting to favor the separate package @minad suggested.

**Comment by oantolin** [comment]
I have tried running the chronicle prototype, but having read it, does it really record the last input? You save the minibuffer contents in a post command hook, doesn't that hook also run after `minibuffer-force-complete-and-exit`, before the exit hook? I would think so, so that you get the candidate again and not the "input".

**Comment by minad** [comment]
> I have not tried running the chronicle prototype, but having read it, does it really record the last input? You save the minibuffer contents in a post command hook, doesn't that hook also run after minibuffer-force-complete-and-exit, before the exit hook? I would think so, so that you get the candidate again and not the "input".

This is what I use in consult and it seems to work with icomplete and selectrum. I trust it based on that, but I am not 100% sure.

**Comment by oantolin** [comment]
> This is what I use in consult and it seems to work with icomplete and selectrum. I trust it based on that, but I am not 100% sure.

But it's tricky then, it depends on the fact that when you press `RET`, the exit hook runs before the post command hook? Is that what happens? Because that sounds weird.

**Comment by oantolin** [comment]
Well, no, maybe that is what happens, and it doesn't so weird...

**Comment by minad** [comment]
> But it's tricky then, it depends on the fact that when you press RET, the exit hook runs before the post command hook? Is that what happens? Because that sounds weird.

No idea tbh, I also found it weird. I found this by experimentation. I cannot give you a sound answer, sorry! Maybe it really is broken and should be replaced by something better in consult too!

**Comment by oantolin** [comment]
Also, I didn't know `consult` had an input history mechanism. 😆

**Comment by minad** [comment]
@clemera @oantolin Regarding the history package, we seem to agree? So the plan is the following:

* We create some separate package handling input histories, named for example chronicle or another name, maybe uchronia if we like weird names
* @oantolin removes embark-repeat, since it is new and not yet used, probably just revert the relevant commit
* @clemera removes selectrum-repeat as soon as some kind of chronicle becomes available
* I will kick out the consult input history mechanism as soon as chronicle becomes available

**Comment by oantolin** [comment]
> No idea tbh, I also find it weird. I found this by experimentation. I cannot give you a sound answer, sorry! Maybe it really is broken and should be replaced by something better in consult too!

No, I just ran some experiments and I think it's a good solution. What actually seems to happen is that the post command hook does not run at all for the command that exits the minibuffer (the one bound to `RET`)! Probably because exiting the minibuffer involves `throw`.

**Comment by oantolin** [comment]
> @clemera @oantolin Regarding the history package, we seem to agree? So the plan is the following:
> 
> * We create some separate package handling input histories, named for example chronicle or another name, maybe uchronia if we like weird names
> * @oantolin removes embark-repeat, since it is new and not yet used, probably just revert the relevant commit
> * @clemera removes selectrum-repeat as soon as some kind of chronicle becomes available
> * I will kick out the consult input history mechanism as soon as chronicle becomes available

I agree with all steps in this plan. And I do like the name `chronicle`, and with @minad's implementation strategy now that I know the post command hook does not run for `RET`.

**Comment by minad** [comment]
I like uchronia now since it is less generic than chronicle and we are basically moving back in the history and rewrite it :rofl: But besides that it looks like we have a plan.

**Comment by oantolin** [comment]
embark-repeat has been reverted. `uchronia` is ok too, but it means a fictional time, no? Maybe there's a name that sounds more like going backwards in time.

**Comment by minad** [comment]
Uchronia is a neologism, but it sounds like a proper word. https://en.wikipedia.org/wiki/Alternate_history
Uchronie apparently is a german word, I certainly didn't know that :rofl: And ucronía is a proper spanish word?

We could also go with ucronia but I would somehow miss the h then...

Wikipedia:
> In Spanish, French, German, Portuguese, Italian, Catalan and Galician, the genre of alternate history is called uchronie / ucronia / ucronía / Uchronie, which has given rise to the term Uchronia in English.

**Comment by clemera** [comment]
I like the suggested names another option that came to mind: DeLorean

**Comment by oantolin** [comment]
I did not know ucronía was a Spanish word! It means "a reconstruction of history based on hypothetical data".

You know, what? I like `uchronia` now. You can argue that rerunning an old command and its input starts an alternate timeline.

Plus, as you say, it's less generic than `chronicle`. And `chronicle` sounds like it should be some journalling package or something.

**Comment by minad** [comment]
I think DeLorean should be reserved for something more sophisticated :rofl:

**Comment by minad** [comment]
> You can argue that rerunning an old command and its input starts an alternate timeline.

This is exactly what I thought!

**Comment by minad** [comment]
I've made a small uchronia mode here: https://github.com/minad/uchronia/blob/main/uchronia.el. It works well, but I am still not 100% convinced that this is the right thing since it is very small, because of the split etc. Sorry for being so undecided...

Now there is an alternative I would like to consider:

* Keep the input history special handling in consult, every command can decide by itself if it wants to use input or not by using consult--read or the technique from there!
* Add consult-repeat, consult-repeat-select and consult-repeat-history which is basically uchronia-repeat, uchronia-select and uchronia-command-history, this will just enhance all existing commands and allow repeating sessions. We could as well put this into Embark if we prefer that!
* Do **not** provide the functionality of the history toggler. This is some functionality I don't like. I don't think it is useful and I would neither like to see it in selectrum/embark/consult. But since this has been discussed in https://github.com/raxod502/selectrum/pull/185, I would like to hear why this has been considered? Maybe @clemera could enlighten me.
* Unfortunately at this point the possibility to use the input history for existing commands has been lost. But this could be restored by replacing history elements with input of existing commands and shouldn't introduce problems.

Basically I just propose a downstripped uchronia without the possibility to toggle between input/normal histories. And I think at this point it is just so small that I would rather put it into consult/embark itself!

**Comment by clemera** [comment]
> Do not provide the functionality of the history toggler. This is some functionality I don't like. I don't think it is useful and I would neither like to see it in selectrum/embark/consult. But since this has been discussed in raxod502/selectrum#185, I would like to hear why this has been considered? Maybe @clemera could enlighten me.

I don't remember exactly, I think it was considered because we thought both can be useful and it is convenient to toggle to the behaviour you are interested in. The candidate history is useful if you are looking for a candidate you used before the input history if you search for search results you previously had.

**Comment by minad** [comment]
@clemera Yes, I agree that it is not totally useless. But still - I don't think the tradeoffs are really worth it. By Emacs standards, there is one history per command and my proposal is to just allow to define it as either input or candidate. I don't like the text property solution, I rather introduced separate variables now in (https://github.com/minad/uchronia/blob/4cf9d72cf673352080fe835324d8f35e2b9ee43f/uchronia.el). I am worried that the text property solution could potentially break other things. But maybe if I would embrace the text property solution, I would feel more convinced than I am now, since then we would at least not get the history variable duplication. But I argue that having only one history type covers like 99% of the use cases. You can always obtain the selected candidate again, by repeating with the input history and then making the selection. And then having a changed set of candidates - I guess this is such a rare use case. In Consult I also only have one history type per command and so do all Emacs commands. Do other packages also provide the functionality to toggle between the history types? All in all I am thinking if I would like to use such a toggler, and from my current feeling, I don't think I would. So it wouldn't pass this kind of check and I don't want to implement functionality of which I am not somehow convinced. Have you been in a situation where you would have wanted that functionality? I have not, but this is always easy to say since with the given restrictions it may also not have occurred to me that I could want something like this.

**Comment by clemera** [comment]
The history commands don't care about the text properties but I agree that you never know in Emacs land, maybe someone relies on the fact the history doesn't has text properties by default but if this becomes relevant you could also remove it then. If you only allow input history for a command you are also dependent on the filtering method used, when you switch the framework or default completion you might still want to be able to access previous candidates.

**Comment by minad** [comment]
@clemera I will try the text property method, maybe I am more satisfied then since it may feel more "low cost", not introducing these auxiliary `uchronia-input-history/...` histories. We will see. At least it is good to do this kind of experimentation outside of Embark and Consult :)
